src/log.rs

```diff
@@ -1,47 +1,152 @@
+    cookie: *mut c_void,
+    callback: Dav1dLoggerCallback,
+}
+
+impl fmt::Write for OnlyDav1dLogger {
+    fn write_str(&mut self, s: &str) -> fmt::Result {
+        // `s` doesn't have a terminating nul-byte,
+        // and it may have internal nul-bytes,
+        // so it's easiest just to print one byte at a time.
+        // This may be slow, but logging can be disabled if it's slow,
+        // or the Rust API can be used instead.
+        let fmt = cstr!("%c");
+        for &byte in s.as_bytes() {
+            // # Safety
+            //
+            // The first argument is `self.cookie`
+            // and the rest are safe to call `printf` with,
+            // as required by [`Self::new`].
+            unsafe {
+                (self.callback)(self.cookie, fmt.as_ptr(), byte as c_uint);
+            }
         }
+        Ok(())
     }
 }
 
-#[cold]
-pub unsafe extern "C" fn rav1d_log(c: *mut Rav1dContext, format: *const c_char, args: ...) {
-    if validate_input!(!c.is_null()).is_err() {
-        return;
+#[derive(Clone, Default)]
+pub(crate) enum Rav1dLogger {
+    Dav1d(OnlyDav1dLogger),
+    Stdout,
+    #[default]
+    Stderr,
+}
+
+/// Any type implementing [`Rav1dLog`] can be used with [`write!`].
+pub trait Rav1dLog {
+    fn write_fmt(&self, args: fmt::Arguments);
+}
+
+impl Rav1dLog for Rav1dLogger {
+    /// Logging doesn't have to be fast when it's on,
+    /// but we don't want it slow things down when it's off,
+    /// so ensure the logging code isn't inlined everywhere, bloating call sites.
+    #[inline(never)]
+    fn write_fmt(&self, args: fmt::Arguments) {
+        match self {
+            // The `dav1d.clone()` is because [`fmt::Write::write_fmt`] takes `&mut`
+            // even though we don't need it to.
+            // [`OnlyDav1dLogger`] is trivial to [`Clone`], though, so we can just do that.
+            Self::Dav1d(dav1d) => dav1d.clone().write_fmt(args).unwrap(),
+            Self::Stdout => stdout().write_fmt(args).unwrap(),
+            Self::Stderr => stderr().write_fmt(args).unwrap(),
+        }
     }
-    if ((*c).logger.callback).is_none() {
-        return;
+}
+
+impl Rav1dLog for Option<Rav1dLogger> {
+    /// When a logger isn't set, we don't want to have to run any code here,
+    /// so force this to be inlined so a [`None`] can be seen.
+    #[inline(always)]
+    fn write_fmt(&self, args: fmt::Arguments) {
+        if let Some(logger) = self {
+            logger.write_fmt(args);
+        }
+    }
+}
+
+/// Used as a marker for [`Rav1dLogger::Stdout`].  Still a valid (i.e. safe) [`Dav1dLoggerCallback`], though.
+unsafe extern "C" fn rav1d_logger_stdout(_cookie: *mut c_void, _fmt: *const c_char, ...) {}
+
+/// Used as a marker for [`Rav1dLogger::Stderr`].  Still a valid (i.e. safe) [`Dav1dLoggerCallback`], though.
+unsafe extern "C" fn rav1d_logger_stderr(_cookie: *mut c_void, _fmt: *const c_char, ...) {}
+
+impl From<Dav1dLogger> for Option<Rav1dLogger> {
```

Comment:
Thinking a bit more on this in the context of the next PR, do we even need the ability to roundtrip between Rav1dLogger and Dav1dLogger? If we're using the library with the legacy C API we have no way to set up stdout or stderr logging without providing a custom callback. C users can only provide a logger in the configuration struct. Can we translate once into the Rust version and never go back?

Reply:
We need to go from `Dav1dLogger` to `Rav1dLogger` when reading a `Dav1dLogger` in `Dav1dSettings` set by the user.  And we need to go from `Rav1dLogger` to `Dav1dLogger` in `fn dav1d_default_settings`, a `DAV1D_API`.  The default is the `stderr` printing.  It's probably not meant to be read, just defaults set and then certain ones overridden, but it's technically part of the ABI.

Reply:
Why do we need to go from a Rav1dLogger to a Dav1dLogger for `dav1d_default_settings`? We could just keep the default log callback (or something similar) and completely separate the two. If you build a logger from C, you get the C callback that prints to stderr, if you build one from Rust you don't get any of the C cruft.

Reply:
We could probably do that, but why?  Right now there's a clear way the `DAV1D_API` is handled everywhere: always convert to `Rav1d*` types and back.  Making exceptions to this pattern increases the cognitive burden of reasoning about any of the `DAV1D_API`s a lot, at least for me.  And since I found a solution here anyways, even if it does some unconventional stuff with the marker `fn` ptrs, what's the point?

Reply:
We discussed this on a call, makes sense for now, even if we don't stick with this forever.